File: lld/week-02-03-patterns/prototype/python/prototype.py
"""Prototype pattern — cloning a Document with nested Metadata instead of
rebuilding it. See ../README.md for the design writeup.
"""
from __future__ import annotations

import copy
from dataclasses import dataclass, field
from typing import Dict, List


@dataclass
class Metadata:
    author: str
    tags: List[str] = field(default_factory=list)


@dataclass
class Document:
    title: str
    meta: Metadata
    sections: List[str] = field(default_factory=list)
    props: Dict[str, str] = field(default_factory=dict)

    def clone(self) -> "Document":
        # copy.deepcopy walks nested Metadata, the sections list, and the
        # props dict, so the clone shares no mutable state with the original.
        return copy.deepcopy(self)


def _demo() -> None:
    original = Document(
        title="Q3 Report",
        meta=Metadata(author="Vinay", tags=["finance", "draft"]),
        sections=["Intro", "Numbers"],
        props={"status": "draft"},
    )

    clone = original.clone()
    clone.meta.author = "Someone Else"
    clone.meta.tags.append("reviewed")
    clone.sections.append("Appendix")
    clone.props["status"] = "final"

    print("Original author:", original.meta.author)
    print("Clone author:", clone.meta.author)
    print("Original tags:", original.meta.tags)
    print("Clone tags:", clone.meta.tags)
    print("Original sections:", original.sections)
    print("Clone sections:", clone.sections)
    print("Original props:", original.props)
    print("Clone props:", clone.props)


if __name__ == "__main__":
    _demo()

File: lld/week-02-03-patterns/prototype/go/prototype.go
// Package prototype implements the Prototype pattern via a Document with
// nested Metadata and section content, cloned via a deep-copy Clone method
// instead of being rebuilt field by field.
package prototype

type Metadata struct {
	Author string
	Tags   []string
}

// clone returns a Metadata with its own backing array for Tags, so mutating
// the clone's tag slice never affects the original.
func (m *Metadata) clone() *Metadata {
	tags := make([]string, len(m.Tags))
	copy(tags, m.Tags)
	return &Metadata{Author: m.Author, Tags: tags}
}

type Document struct {
	Title    string
	Meta     *Metadata
	Sections []string
	Props    map[string]string
}

// Clone performs a deep copy: the nested Metadata, the Sections slice, and
// the Props map are all duplicated, not shared, with the original.
func (d *Document) Clone() *Document {
	sections := make([]string, len(d.Sections))
	copy(sections, d.Sections)

	props := make(map[string]string, len(d.Props))
	for k, v := range d.Props {
		props[k] = v
	}

	return &Document{
		Title:    d.Title,
		Meta:     d.Meta.clone(),
		Sections: sections,
		Props:    props,
	}
}
